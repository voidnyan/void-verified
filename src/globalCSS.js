import { StyleHandler } from "./styleHandler";

export class GlobalCSS {
	settings;
	styleHandler;

	styleId = "global-css";
	isCleared = false;

	cssInLocalStorage = "void-verified-global-css";
	constructor(settings) {
		this.settings = settings;
		this.styleHandler = new StyleHandler(settings);

		this.css = localStorage.getItem(this.cssInLocalStorage) ?? "";
	}

	createCss() {
		if (!this.settings.options.globalCssEnabled.getValue()) {
			this.styleHandler.clearStyles(this.styleId);
			return;
		}

		if (!this.shouldRender()) {
			return;
		}

		this.isCleared = false;
		this.styleHandler.createStyleLink(this.css, this.styleId);
	}

	updateCss(css) {
		this.css = css;
		this.createCss();
		localStorage.setItem(this.cssInLocalStorage, css);
	}

	clearCssForProfile() {
		if (this.isCleared) {
			return;
		}
		if (!this.shouldRender()) {
			this.styleHandler.clearStyles(this.styleId);
			this.isCleared = true;
		}
	}

	shouldRender() {
		if (window.location.pathname.startsWith("/settings")) {
			return false;
		}

		if (!this.settings.options.globalCssAutoDisable.getValue()) {
			return true;
		}

		if (
			!window.location.pathname.startsWith("/user/") &&
			!window.location.pathname.startsWith("/activity/")
		) {
			return true;
		}

		const profileCustomCss = document.getElementById(
			"customCSS-automail-styles"
		);

		const styleHandler = new StyleHandler(this.settings);
		const voidActivityStyles = styleHandler.getStyleLink("activity-css");
		const voidUserStyles = styleHandler.getStyleLink("user-css");

		if (voidActivityStyles || voidUserStyles) {
			return false;
		}

		if (!profileCustomCss) {
			return true;
		}

		const shouldRender = profileCustomCss.innerHTML.trim().length === 0;
		return shouldRender;
	}
}
