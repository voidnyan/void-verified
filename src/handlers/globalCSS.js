import { StyleHandler } from "./styleHandler";
import {LocalStorageKeys} from "../assets/localStorageKeys";

export class GlobalCSS {
	settings;
	styleHandler;

	styleId = "global-css";
	isCleared = false;

	cssInLocalStorage = LocalStorageKeys.globalCSS;
	constructor(settings) {
		this.settings = settings;
		this.styleHandler = new StyleHandler(settings);

		this.css = localStorage.getItem(this.cssInLocalStorage) ?? "";
	}

	createCss() {
		if (!this.settings.options.globalCssEnabled.getValue()) {
			this.styleHandler.clearStyles(this.styleId);
			return;
		}

		this.isCleared = false;
		this.styleHandler.createStyleLink(this.css, this.styleId);
	}

	updateCss(css) {
		this.css = css;
		this.createCss();
		localStorage.setItem(this.cssInLocalStorage, css);
	}

}
