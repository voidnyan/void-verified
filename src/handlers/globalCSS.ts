import { StyleHandler } from "./styleHandler";
import {LocalStorageKeys} from "../assets/localStorageKeys";
import {StaticSettings} from "../utils/staticSettings";
import {DOM} from "../utils/DOM";
import {AceEditorInitializer} from "../utils/aceEditorInitializer";
import {Button} from "../components/components";

export class GlobalCSS {
	static styleHandler;
	static styleId = "global-css";
	static css: string;

	static initialize() {
		this.styleHandler = new StyleHandler(StaticSettings.settingsInstance);
		this.css = localStorage.getItem(LocalStorageKeys.globalCSS) ?? "";
		this.createCss();
	}

	static createCss() {
		if (!StaticSettings.options.globalCssEnabled.getValue() || window.location.pathname.startsWith("/settings/developer")) {
			this.styleHandler.clearStyles(this.styleId);
			return;
		}

		this.styleHandler.createStyleLink(this.css, this.styleId);
	}

	static updateCss(css) {
		this.css = css;
		this.createCss();
		localStorage.setItem(LocalStorageKeys.globalCSS, css);
	}


	static renderEditor() {
		const cssName = "global";
		const container = DOM.createDiv("css-editor");
		const label = DOM.create("h3", null, `Custom ${cssName} CSS`);
		container.append(label);


		const editor = AceEditorInitializer.createEditor(`custom-css-editor-${cssName}`, this.css);
		container.append(editor);
		AceEditorInitializer.addChangeHandler(`custom-css-editor-${cssName}`, (value) => {
			this.updateCss(value);
		});

		if (cssName === "global") {
			const notice = DOM.create("div");
			notice.innerText =
				"Please note that Custom CSS is disabled in the settings. \nIn the event that you accidentally disable rendering of critical parts of AniList, navigate to the settings by URL";
			notice.style.fontSize = "11px";
			container.append(notice);
		}

		const prettifyButton = Button("Prettify", () => {
			// @ts-ignore
			const beautify = ace.require("ace/ext/beautify");
			// @ts-ignore
			const editor = ace.edit(`void-custom-css-editor-${cssName}`);
			const value = editor.getValue()
				.replace(/(\n\s*\n)+/g, '\n\n')
				.replace(/\{[^\}]*\}/g, (block) => {
					// Remove all empty lines within the block
					return block.replace(/\n\s*\n/g, '\n');
				});
			editor.setValue(value);
			beautify.beautify(editor.session);
		});
		container.append(prettifyButton);
		return container;
	}

}
