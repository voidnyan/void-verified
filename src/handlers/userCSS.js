import {AnilistAPI} from "../api/anilistAPI";
import {ColorFunctions} from "../utils/colorFunctions";
import {StyleHandler} from "./styleHandler";
import LZString from "../libraries/lz-string";
import {Toaster} from "../utils/toaster";
import {DOM} from "../utils/DOM";
import {ActionInputField, Link, TextArea} from "../components/components";
import {SearchDocumentIcon} from "../assets/icons";
import {AceEditorInitializer} from "../utils/aceEditorInitializer";
import {CssCache} from "../utils/cssCache";

export class UserCSS {
	#settings;
	#currentActivity;
	#currentUser;
	css = "";
	preview = false;
	cssInLocalStorage = "void-verified-user-css";
	broadcastChannel;
	#csspy = {
		username: "",
		css: "",
	};

	constructor(settings) {
		this.#settings = settings;
		if (
			this.#settings.auth?.token &&
			this.#settings.options.profileCssEnabled.getValue()
		) {
			const cssInLocalStorage = JSON.parse(
				localStorage.getItem(this.cssInLocalStorage),
			);
			if (cssInLocalStorage) {
				this.css = cssInLocalStorage.css;
				this.preview = cssInLocalStorage.preview;
			} else {
				this.getAuthUserCss();
			}
		}

		this.broadcastChannel = new BroadcastChannel("user-css");
		this.broadcastChannel.addEventListener("message", (event) =>
			this.#handleBroadcastMessage(event, this.#settings),
		);
	}

	async checkActivityCss() {
		if (
			!this.#settings.options.activityCssEnabled.getValue() ||
			!window.location.pathname.startsWith("/activity/")
		) {
			return;
		}

		const activityId = window.location.pathname.match(
			/^\/activity\/([^/]*)\/?/,
		)[1];

		if (this.#currentActivity === activityId) {
			return;
		}

		this.#currentActivity = activityId;

		if (this.#settings.options.cacheUserCss.getValue() && !this.preview) {
			const activityUsername = document.querySelector(".activity-entry > .wrap a.name").innerText.trim();
			if (!this.#shouldRenderCss(activityUsername)) {
				return;
			}
			const cachedCss = CssCache.get(activityUsername);
			if (cachedCss) {
				new StyleHandler(this.#settings).clearStyles("user-css");
				this.#renderCss(cachedCss.css, "user-css");
				this.#setActivityColor(cachedCss.userColor);
				Toaster.debug("Found CSS from cache.");
				return;
			}
		}

		let activity;
		try {
			Toaster.debug("Querying user activity.");
			const anilistAPI = new AnilistAPI(this.#settings);
			activity = await anilistAPI.getActivityCss(activityId);
		} catch {
			Toaster.error("Failed to get activity CSS.");
			return;
		}

		const username = activity.user?.name ?? activity.recipient?.name;

		const userColor =
			activity.user?.options.profileColor ??
			activity.recipient?.options.profileColor;

		this.#setActivityColor(userColor);

		if (username === this.#settings.anilistUser && this.preview) {
			this.#renderCss(this.css, "user-css");
			return;
		}

		if (username === this.#currentUser) {
			this.#clearGlobalCss();
			return;
		}
		new StyleHandler(this.#settings).clearStyles("user-css");

		if (!this.#shouldRenderCss(username)) {
			return;
		}

		const about = activity.user?.about ?? activity.recipient?.about;

		const css = this.#decodeAbout(about)?.customCSS;
		if (css) {
			this.#renderCss(css, "user-css");
		} else {
			Toaster.debug("User has no custom CSS.");
		}

		this.#currentUser = username;

		if (this.#settings.options.cacheUserCss.getValue() && (css || userColor)) {
			CssCache.set(username, css, userColor);
		}
	}

	#setActivityColor(userColor) {
		const rgb = ColorFunctions.handleAnilistColor(userColor);

		const activityEntry = document.querySelector(
			".container > .activity-entry",
		);

		activityEntry.style.setProperty("--color-blue", rgb);
		activityEntry.style.setProperty("--color-blue-dim", rgb);
	}

	resetCurrentActivity() {
		this.#currentActivity = null;
	}

	async checkUserCss() {
		if (
			!this.#settings.options.profileCssEnabled.getValue() ||
			!window.location.pathname.startsWith("/user/")
		) {
			return;
		}

		const username =
			window.location.pathname.match(/^\/user\/([^/]*)\/?/)[1];

		if (username === this.#currentUser) {
			return;
		}

		if (username === this.#settings.anilistUser && this.preview) {
			this.#renderCss(this.css, "user-css");
			return;
		}

		if (!this.#shouldRenderCss(username)) {
			new StyleHandler(this.#settings).clearStyles("user-css");
			return;
		}

		this.#currentUser = username;

		if (this.#settings.options.cacheUserCss.getValue()) {
			const cachedCss = CssCache.getCss(username);
			if (cachedCss) {
				Toaster.debug("Found CSS from cache.");
				this.#renderCss(cachedCss, "user-css");
				return;
			}
		}

		let about;
		let userColor;
		try {
			Toaster.debug("Querying user CSS.");
			const anilistAPI = new AnilistAPI(this.#settings);
			const user = await anilistAPI.getUserCssAndColour(username)
			about = user?.about;
			userColor = user?.options?.profileColor;
		} catch (error) {
			Toaster.error("Failed to load user's CSS.");
			return;
		}

		const css = this.#decodeAbout(about)?.customCSS;
		if (!css) {
			Toaster.debug("User has no custom CSS.");
			new StyleHandler(this.#settings).clearStyles("user-css");
		}
		this.#renderCss(css, "user-css");
		if (this.#settings.options.cacheUserCss.getValue() && (css || userColor)) {
			CssCache.set(username, css, userColor);
		}
	}

	resetCurrentUser() {
		this.#currentUser = null;
	}

	updateCss(css) {
		this.css = css;
		if (this.preview) {
			this.broadcastChannel.postMessage({type: "css", css});
		}
		this.#saveToLocalStorage();
	}

	async publishUserCss() {
		const username = this.#settings.anilistUser;
		if (!username) {
			return;
		}

		const anilistAPI = new AnilistAPI(this.#settings);
		let about;
		try {
			Toaster.debug("Querying account user about to merge changes into.");
			about = await anilistAPI.getUserAbout(username);
		} catch (error) {
			Toaster.error("Failed to get current about for merging new CSS.");
			return;
		}
		if (!about) {
			about = "";
		}
		let aboutJson = this.#decodeAbout(about);
		aboutJson.customCSS = this.css;
		const compressedAbout = LZString.compressToBase64(
			JSON.stringify(aboutJson),
		);

		const target = about.match(/^\[\]\(json([A-Za-z0-9+/=]+)\)/)?.[1];

		if (target) {
			about = about.replace(target, compressedAbout);
		} else {
			about = `[](json${compressedAbout})\n\n` + about;
		}
		try {
			Toaster.debug("Publishing CSS.");
			await anilistAPI.saveUserAbout(about);
			if (this.#settings.options.cacheUserCss.getValue()) {
				CssCache.clear(this.#settings.anilistUser);
			}
			Toaster.success("CSS published.");
		} catch (error) {
			Toaster.error("Failed to publish CSS changes.");
		}
	}

	togglePreview() {
		this.preview = !this.preview;
		this.broadcastChannel.postMessage({
			type: "preview",
			preview: this.preview,
		});
		this.#saveToLocalStorage();
	}

	async getAuthUserCss() {
		const anilistAPI = new AnilistAPI(this.#settings);
		const username = this.#settings.anilistUser;
		if (!username) {
			return;
		}
		try {
			Toaster.debug("Querying account user CSS.");
			const about = await anilistAPI.getUserAbout(username);
			const css = this.#decodeAbout(about).customCSS;
			this.css = css;
			this.#saveToLocalStorage();
			return css;
		} catch (error) {
			Toaster.error("Failed to query account user CSS.");
		}
	}

	renderCSSpy(settingsUi) {
		const container = DOM.create("div");
		container.append(DOM.create("h3", null, "CSSpy"));

		const usernameInput = ActionInputField(
			"",
			(_, inputField) => {
				this.#handleSpy(inputField, settingsUi);
			},
			SearchDocumentIcon(),
		);

		container.append(usernameInput);

		if (this.#csspy.css === "") {
			return container;
		}

		const header = DOM.create("h5", "layout-header", [
			Link(
				this.#csspy.username,
				`https://anilist.co/user/${this.#csspy.username}/`,
			),
			`'s CSS`,
		]);

		const cssContainer = AceEditorInitializer.createEditor("css-spy-container", this.#csspy.css);

		setTimeout(() => {
			ace.edit("void-css-spy-container").setOption("readOnly", true);
		}, 160);

		container.append(header);
		container.append(cssContainer);

		return container;
	}

	async #handleSpy(inputField, settingsUi) {
		const username = inputField.value;
		if (username === "") {
			return;
		}

		try {
			Toaster.debug(`Spying CSS from ${username}.`);
			const about = await new AnilistAPI(this.#settings).getUserAbout(
				username,
			);
			const css = this.#decodeAbout(about)?.customCSS;
			if (css) {
				this.#csspy.css = css;
				this.#csspy.username = username;
				settingsUi.renderSettingsUiContent();
			} else {
				this.#csspy.css = "";
				this.#csspy.username = "";
				Toaster.debug("User has no custom CSS.");
			}
		} catch (error) {
			this.#csspy.css = "";
			this.#csspy.username = "";
			Toaster.error(`There was an error getting CSS for ${username}`);
			console.error(error);
		}
	}

	#handleBroadcastMessage(event, settings) {
		switch (event.data.type) {
			case "css":
				this.#handlePreviewCssMessage(event.data.css, settings);
				break;
			case "preview":
				this.#handlePreviewToggleMessage(event.data.preview);
				break;
		}
	}

	#handlePreviewCssMessage(css, settings) {
		this.css = css;
		const hasUserCss = document.getElementById(
			"void-verified-user-css-styles",
		);
		if (hasUserCss) {
			new StyleHandler(settings).createStyleLink(css, "user-css");
		}
	}

	#handlePreviewToggleMessage(preview) {
		this.preview = preview;
		const hasUserCss = document.getElementById(
			"void-verified-user-css-styles",
		);
		if (!hasUserCss) {
			return;
		}

		this.resetCurrentUser();
		this.resetCurrentActivity();

		this.checkUserCss();
		this.checkActivityCss();
	}

	#saveToLocalStorage() {
		localStorage.setItem(
			this.cssInLocalStorage,
			JSON.stringify({
				css: this.css,
				preview: this.preview,
			}),
		);
	}

	#shouldRenderCss(username) {
		const user = this.#settings.getUser(username);
		if (
			this.#settings.options.onlyLoadCssFromVerifiedUser.getValue() &&
			!this.#settings.isVerified(username)
		) {
			return false;
		}
		if (user?.onlyLoadCssFromVerifiedUser) {
			return true;
		}
		return !this.#userSpecificRenderingExists();
	}

	#userSpecificRenderingExists() {
		return this.#settings.verifiedUsers.some(
			(user) => user.onlyLoadCssFromVerifiedUser,
		);
	}

	#renderCss(css, id) {
		if (!css) {
			return;
		}

		const styleHandler = new StyleHandler(this.#settings);
		styleHandler.createStyleLink(css, id);
		this.#clearGlobalCss();
	}

	#clearGlobalCss() {
		if (this.#settings.options.globalCssAutoDisable.getValue()) {
			new StyleHandler(this.#settings).clearStyles("global-css");
		}
	}

	#decodeAbout(about) {
		let json = (about || "").match(/^\[\]\(json([A-Za-z0-9+/=]+)\)/);
		if (!json) {
			return {
				customCss: "",
			};
		}

		let jsonData;
		try {
			jsonData = JSON.parse(atob(json[1]));
		} catch (e) {
			jsonData = JSON.parse(LZString.decompressFromBase64(json[1]));
		}
		return jsonData;
	}
}
