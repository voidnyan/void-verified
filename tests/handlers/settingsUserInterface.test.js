import { Settings } from "../../src/utils/settings";
import { SettingsUserInterface } from "../../src/handlers/settingsUserInterface";
import { StyleHandler } from "../../src/handlers/styleHandler";
import { GlobalCSS } from "../../src/handlers/globalCSS";
import { UserCSS } from "../../src/handlers/userCSS";
import { LayoutDesigner } from "../../src/handlers/layoutDesigner";

beforeAll(() => {
	jest.spyOn(global.Math, "random").mockReturnValue(0.12352);
});

afterAll(() => {
	jest.spyOn(global.Math, "random").mockRestore();
});

describe("Settings User Interface", () => {
	it("renders", () => {
		const settings = new Settings();
		const styleHandler = new StyleHandler(settings);
		const globalCSS = new GlobalCSS(settings);
		const userCSS = new UserCSS(settings);
		const layoutDesigner = new LayoutDesigner(settings);

		const settingsUserInterface = new SettingsUserInterface(
			settings,
			styleHandler,
			globalCSS,
			userCSS,
			layoutDesigner
		);

		const settingsContainer = document.createElement("div");
		settingsContainer.classList.add("settings");
		settingsContainer.classList.add("container");

		const content = document.createElement("div");
		content.classList.add("content");
		settingsContainer.append(content);

		document.body.append(settingsContainer);
		settingsUserInterface.renderSettingsUi();

		const settingsUserInterfaceDOM = document.querySelector(
			"#void-verified-settings"
		);

		expect(settingsUserInterfaceDOM).not.toBeNull();
		expect(settingsUserInterfaceDOM).toMatchSnapshot();
	});
});
