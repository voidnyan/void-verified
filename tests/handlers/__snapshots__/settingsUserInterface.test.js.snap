// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings User Interface renders 1`] = `
<div
  class="void-settings"
  id="void-verified-settings"
>
  <div>
    <div
      class="void-settings-header"
    >
      <h1>
        VoidVerified Settings
      </h1>
      <p>
        Version: 
        <span>
          TEST
        </span>
      </p>
      <p>
        Author: 
        <a
          class="void-link"
          href="https://anilist.co/user/voidnyan/"
          target="_blank"
        >
          voidnyan
        </a>
      </p>
      <button
        class="void-button"
      >
        View Changelog
      </button>
    </div>
    <nav
      class="void-nav"
    >
      <ol>
        <li
          class="void-active"
        >
          all
        </li>
        <li>
          users
        </li>
        <li>
          paste
        </li>
        <li>
          css
        </li>
        <li>
          misc
        </li>
      </ol>
    </nav>
    <div
      class="void-settings-list"
    >
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Copy user color from their profile.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Move activity subscribe button next to comments and likes.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Hide activity and reply like counts.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display a verified sign next to usernames.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display a verified sign next to a profile name.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-input"
          id="0.12352"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          The default sign displayed next to a username.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Highlight user activity with a border.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Highlight replies with a border.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-input"
          id="0.12352"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Width of the highlight border.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Color user activity links with user color.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Color user reply links with user color.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Use fallback highlight color when user color is not specified.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Enable custom global CSS.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Disable global CSS when a profile has custom CSS.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Load user's custom CSS when viewing their profile.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Load user's custom CSS when viewing their activity (direct link).
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Only load custom CSS from verified users.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Enable Layout Designer in the settings tab.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display quick access of users in home page.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display quick access of notifications in home page.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display a badge on quick access when changes are detected on user's layout.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display a timer until next update of Quick Access.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Automatically wrap pasted links and images with link and image tags.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Wrap images with a link tag.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-input"
          id="0.12352"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Width used when pasting images.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Upload image from the clipboard to image host (configure below).
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Enable toast notifications.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Query AniList API with elevated browser access (this might solve some API issues).
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Open AniList links in the same tab.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Add a GIF keyboard to activity editor.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Add like buttons to add media to GIF keyboard.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Display a changelog when a new version is detected.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Enable a self-message button on your profile (requires authentication).
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Fix AniList bug where private messages are displayed in List activity feed.
        </label>
      </div>
      <div
        class="void-setting-label-container"
      >
        <input
          class="void-checkbox"
          id="0.12352"
          type="checkbox"
        />
        <label
          class="void-setting-label"
          for="0.12352"
        >
          Enable CSSpy in Layout & CSS tab.
        </label>
      </div>
    </div>
    <nav
      class="void-nav"
    >
      <ol>
        <li
          class="void-active"
        >
          users
        </li>
        <li>
          authorization
        </li>
        <li>
          toasts
        </li>
      </ol>
    </nav>
    <div
      class="void-table"
      id="void-verified-user-table"
      style="margin-top: 25px;"
    >
      <table
        class="void-table"
      >
        <thead>
          <tr>
            <th>
              Username
            </th>
            <th>
              Sign
            </th>
            <th>
              Color
            </th>
            <th>
              Other
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
      <form>
        <label
          for="void-verified-add-user"
        >
          Add user
        </label>
        <input
          class="void-input"
          id="void-verified-add-user"
        />
      </form>
    </div>
    <h5>
      Fallback color
    </h5>
    <div
      class="void-color-picker-container"
    >
      <input
        class="void-color-picker"
        type="color"
      />
      <input
        class="void-color-picker-input"
        type="text"
      />
    </div>
  </div>
</div>
`;
